### ISSUE TITLE
Grad-CAM and Grayscale-CAM Results Remain Unchanged Despite Trained Discriminator

### ISSUE THREAD

**Issue by iremalti** [jacobgil/pytorch-grad-cam#444]
Hello,

I am facing an issue with the Grad-CAM visualization of my model, which acts as a discriminator in a GAN. I attempted to compute Grad-CAM for 50 images from a dataset, but the results are identical for all images and remain unchanged, even though the discriminator has undergone training and improvement. Additionally, I found that the Grayscale-CAM values for all images consistently remain at 0 throughout the entire training process.

Expected Behavior:
The Grad-CAM results should change during the training process as the discriminator improves, and the discriminating features for real and fake images should vary.

model = discriminator
 target_layers = [model.main[-2]]
 targets = [BinaryClassifierOutputTarget(0)]

Discriminator:

 class Diskriminator(nn.Module):
     def __init__(self):
         super(Diskriminator, self).__init__()
         self.main = nn.Sequential( nn.Conv2d(IMAGE_CHANNEL, D_HIDDEN, 4, 2, 1, bias=False), 
            nn.LeakyReLU(0.2, inplace=False), 
            nn.Conv2d(D_HIDDEN, D_HIDDEN * 2, 4, 2, 1, bias=False),
             nn.BatchNorm2d(D_HIDDEN * 2), 
            nn.LeakyReLU(0.2, inplace=False), 
            nn.Conv2d(D_HIDDEN * 2, D_HIDDEN * 4, 4, 2, 1, bias=False), 
            nn.BatchNorm2d(D_HIDDEN * 4),
             nn.LeakyReLU(0.2, inplace=False), 
            nn.Conv2d(D_HIDDEN * 4, D_HIDDEN * 8, 4, 2, 1, bias=False),
             nn.BatchNorm2d(D_HIDDEN * 8), 
            nn.LeakyReLU(0.2, inplace=False),
             nn.Conv2d(D_HIDDEN * 8, 1, 4, 1, 0, bias=False), 
            nn.Sigmoid()         )  

    def forward(self, x): 
        x = self.main(x).view(-1, 1).squeeze(1)
         return x

CODE:

for batch in dataloader:     
  images = batch[0].to(device)      
  for image in images:        
     bilder.append(image)         
    count += 1          
    if count >= 50:             break      
  if count >= 50:         break
bilder_input = torch.stack(bilder) 

# GAN training 
for epoch in range (EPOCH_NUM):
...

    if i % 100 == 0: 
          model = discriminator
           target_layers = [model.main[-2]]
           targets = [BinaryClassifierOutputTarget(0)]

          
          for k in range (bilder_input.size(0)):     
              image = bilder_input[k].cpu().numpy() 
              img = image.transpose((1, 2, 0)) 
               org = np.clip(img, 0, 1)  
              org = cv2.cvtColor(org, cv2.COLOR_BGR2RGB)  
              org = (org * 255).astype(np.uint8)  
              cv2.imwrite(os.path.join(output_folder_real, 'Bild_real{'+str(epoch)+'}_{'+str(i)+'}_{'+str(k)+'}_cam.jpg'), org)    
          
              with GradCAM(model=model, target_layers=target_layers, use_cuda=torch.cuda.is_available()) as cam:  
                  grayscale_cam = cam(input_tensor=bilder_input,  targets=targets, aug_smooth=False,
 eigen_smooth=False)  
                  grayscale_cam = grayscale_cam[k, :] 
                   cam_image = show_cam_on_image(img, grayscale_cam, use_rgb=True)
                  cam_image = cv2.cvtColor(cam_image, cv2.COLOR_RGB2BGR)    
              cv2.imwrite(os.path.join(output_folder_real, 'Bild_real{'+str(epoch)+'}_{'+str(i)+'}_{'+str(k)+'}_cam.jpg'), cam_image) 

I would appreciate any assistance in diagnosing the issue with the Grad-CAM results and welcome any suggested solutions or further steps for improvement.

Thank you
